Serde blanket implementation feature

stalkopat opened the pull request:
When working with types in outside crates it quickly becomes a nuisance to implement various wrapper types just to be able to implement BytesConvertable. This becomes especially annoying once one has to interface with an existing codebase that has the original types deeply ingrained.

This PR adds an optional feature called "blanket_serde" which replaces the current default implementations of BytesConvertable with a blanket implementation utilizing serde and pot.

In the process a handful of bugs were found, where tests and BytesConvertable implementations were inconsistent in their Deserialization / Serialization.
For example there were cases of Strings into_bytes() function being used for serialization in combination with <String as BytesConvertable>::from_bytes being used for deserialization, this didn't cause any errors since the default implementation similarly just called .into_bytes().

slawlor commented on ractor/src/serialization.rs:
```diff
@@ -24,7 +24,8 @@ pub trait BytesConvertable {
 #[cfg(feature = "blanket_serde")]
-mod impls{
+/// Contains a blanket implementation for all types that implement serde::Serialize and serde::Deserialize
+pub mod impls{
```
So a user needs to manually

```rust
use ractor::serialization::impls::*;
```

in order to get the implementations? I feel like this should be a `pub use impls::*` somewhere so they get the blanket implementations for free.

stalkopat replied:
No, apparently this module doesn't even have to be public for the blanket implementation to be available, for some reason I vaguely remember not receiving the blanket implementation without making the module public, but can't seem to reproduce it now.
Activating the "cluster" and "blanket_serde" feature is enough on its own to receive the blanket implementations on all the types.

slawlor commented on ractor/src/factory/job.rs:
```diff
@@ -83,8 +83,8 @@ impl BytesConvertable for JobOptions {
 
-            let submit_time = <u64 as BytesConvertable>::from_bytes(data);
-            let ttl = <u64 as BytesConvertable>::from_bytes(ttl_bytes);
+            let submit_time = u64::from_be_bytes(data.try_into().unwrap()); //Unwrap should be safe since we checked length earlier
```
why?

stalkopat replied:
This was inconsistent on serialization / deserialization. the into_bytes() part of JobOptions uses ".to_be_bytes();" on the u64's, while this part fell back to the BytesConvertable implementation of u64, this obviously fails if the underlying BytesConvertable implementation changes (for example this switch to serde or even something more simple like length prefixing the encoded data). Alternative would have been to switch the encoding to use the BytesConvertible variant, but this means we can't make assumptions regarding the offsets during this decoding (namely it being 8 bytes long).

slawlor reviewed (commented):
It's very close to being solid but

1. I want to see the signal results and
2. I think there's some ergonomics possible with the sub-module `impls` for users.

stalkopat commented on .github/workflows/ci.yaml:
```diff
@@ -23,6 +23,9 @@ jobs:
           - name: Test ractor with the `cluster` feature
             package: ractor
             flags: -F cluster
+          - name: Test ractor with the `blanket_serde` feature
```
@slawlor 
Not sure how CI works, presumably this works?

slawlor replied:
No you need to add the feature I think comma separated. There's a matrix of tests to run.

Also since this affects cluster we should run it in the docker suite integration tests

stalkopat commented:
Added blanket_serde to the docker integration tests
Changed the docker integration tests to actually use BytesConvertable
Fixed a bug where the pg_group integration test wrongly returned success when the underlying actor paniced / errored

slawlor reviewed (approved):
Thanks for the contribution! This looks good to merge to me
